File: swirly-elm/src/main/java/com/swirlycloud/domain/Conv.java
/*******************************************************************************
 * Copyright (C) 2013, 2014 Swirly Cloud Limited. All rights reserved.
 *******************************************************************************/
package com.swirlycloud.domain;

public final class Conv {
    private Conv() {
    }

    public static double fractToReal(int numer, int denom) {
        return (double) numer / (double) denom;
    }

    public static long realToIncs(double real, double inc_size) {
        return Math.round(real / inc_size);
    }

    public static double incsToReal(long incs, double inc_size) {
        return incs * inc_size;
    }

    /**
     * Convert quantity to lots.
     */

    public static long qtyToLots(double qty, double qty_inc) {
        return realToIncs(qty, qty_inc);
    }

    /**
     * Convert lots to quantity.
     */

    public static double lotsToQty(long lots, double qty_inc) {
        return incsToReal(lots, qty_inc);
    }

    /**
     * Convert price to ticks.
     */

    public static long priceToTicks(double price, double price_inc) {
        return realToIncs(price, price_inc);
    }

    /**
     * Convert ticks to price.
     */

    public static double ticksToPrice(long ticks, double price_inc) {
        return incsToReal(ticks, price_inc);
    }

    /**
     * Number of decimal places in real.
     */

    public static int realToDp(double d) {
        int dp;
        for (dp = 0; dp < 9; ++dp) {
            final double fp = d % 1.0;
            if (fp < 0.000000001) {
                break;
            }
            d *= 10;
        }
        return dp;
    }

    /**
     * Decimal places as real.
     */

    public static double dpToReal(int dp) {
        return Math.pow(10, -dp);
    }
}

File: swirly-front/src/main/react/conv.js
/*******************************************************************************
 * Copyright (C) 2013, 2015 Swirly Cloud Limited. All rights reserved.
 *******************************************************************************/

function fractToReal(numer, denom) {
    return numer / denom;
}

function realToIncs(real, incSize) {
    return roundHalfAway(real / incSize);
}

function incsToReal(incs, incSize) {
    return incs * incSize;
}

function qtyToLots(qty, contr) {
    return realToIncs(qty, contr.qtyInc);
}

function lotsToQty(lots, contr) {
    return incsToReal(lots, contr.qtyInc).toFixed(contr.qtyDp);
}

function priceToTicks(price, contr) {
    return realToIncs(price, contr.priceInc);
}

function ticksToPrice(ticks, contr) {
    return incsToReal(ticks, contr.priceInc).toFixed(contr.priceDp);
}

function realToDp(d) {
    var dp = 0;
    for (; dp < 9; ++dp) {
        var fp = d % 1.0;
        if (fp < 0.000000001) {
            break;
        }
        d *= 10;
    }
    return dp;
}

function toDateInt(s) {
    return parseInt(s.substr(0, 4) + s.substr(5, 2) + s.substr(8, 2));
}

function toDateStr(i) {
    var year = Math.floor(i / 10000);
    var mon = Math.floor(i / 100) % 100;
    var mday = i % 100;
    return ('000' + year).slice(-4)
        + '-' + ('0' + mon).slice(-2)
        + '-' + ('0' + mday).slice(-2);
}

function toTimeStr(ms) {
    var date = new Date(ms);
    return ('0' + date.getHours()).slice(-2)
        + ':' + ('0' + date.getMinutes()).slice(-2)
        + ':' + ('0' + date.getSeconds()).slice(-2);
}
